### PR TITLE
[IMP] core: read_group: allow grouping by non-stored related fields

### PR DESCRIPTION
**Pull request by yelizariev**
In most cases, the only reason to mark a related field as stored is a need to
group records by such field. With this commit we can make grouping without
making column copy. It allows to save database size as well as resources needed
to create and keep such duplicate up-to-date.

In fact, read_group supported non-stored related fields for ages, but it was
limited to automatic related fields (inherits/delegate=True). So we only need to
weaken constrain for groupby argument and add some extra joinings in
_inherits_join_calc to give new feature to the world.

---

task-2155109

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by randall-vx** [comment]
Amazing job @yelizariev !

File: odoo/addons/test_read_group/models.py
Comment (rco-odoo) [inline odoo/addons/test_read_group/models.py:82]: Use `_inherits = ...` on the model instead.

File: odoo/addons/test_read_group/tests/test_related.py
Comment (rco-odoo) [inline odoo/addons/test_read_group/tests/test_related.py:42]: Maybe make it more explicit.

File: odoo/addons/test_read_group/tests/test_related.py
Comment (yelizariev) [inline odoo/addons/test_read_group/tests/test_related.py:27]: More tests:

*  at least two states
* test against Null values too

File: odoo/addons/test_read_group/tests/test_related.py
Comment (rco-odoo) [inline odoo/addons/test_read_group/tests/test_related.py:25]: The test data should include at least two non-null values and a null (coverage).

File: odoo/models.py
Comment (rco-odoo) [inline odoo/models.py:2441]: I don't think the boolean flag is actually necessary.  The method could handle related fields in all cases.

File: odoo/models.py
Comment (rco-odoo) [inline odoo/models.py:2477]: Not all conditions should be on `field`.
In the case of `country_code`, the condition should be `country_code.compute_sudo or `country_id.auto_join`.

File: odoo/addons/test_read_group/tests/test_related.py
Comment (rco-odoo) [inline odoo/addons/test_read_group/tests/test_related.py:38]: Maybe add a test where `read_group()` fails because of incorrect argument (non-stored related field with `compute_sudo=False` and non-autojoin relation).

File: odoo/addons/test_read_group/models.py
Comment (yelizariev) [inline odoo/addons/test_read_group/models.py:96]: It's needed some more tests with auto_join=True, compute_sudo=False

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2470]: TODO: try to combine process_related and process_inherited

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2490]: @rco-odoo What do you think about this TODO? If it's may be useful, shall I implement it in this PR or leave for now?

**Comment by yelizariev** [comment]
@rco-odoo please review new tests. Once it's approved I'll continue with improving updates in ``_inherits_join_calc``

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2468]: @rco-odoo What do you think about this TODO? If it's may be useful, shall I implement it in this PR or leave for now?

File: odoo/addons/test_read_group/tests/test_related.py
Comment (yelizariev) [inline odoo/addons/test_read_group/tests/test_related.py:68]: The two fields ``state_code`` and ``state_code2`` look the same and are essentially related to  ``partner_id.state_id.code``, but ``state_code2`` is passed via intermediate field ``state_id`` in model ``test_read_group.related2`` and implicitly gets compute_sudo

```
class Related2(models.Model):
    _name = 'test_read_group.related2'
    _inherits = {'test_read_group.related1': 'related1_id'}
    _description = 'related2'

    related1_id = fields.Many2one('test_read_group.related1', required=True)
    state_id = fields.Many2one('res.country.state', related='partner_id.state_id')
```

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2442]: Shall I make the renaming?

File: odoo/addons/test_read_group/tests/test_related.py
Comment (yelizariev) [reply odoo/addons/test_read_group/tests/test_related.py]: I updated first implementation and ``state_code`` doesn't raise access error, because it has another field with auto_join=True

**Comment by yelizariev** [comment]
Related
https://github.com/odoo/odoo/commit/22b0142d4e9d72b80f50388d17ebf3e95d76deba

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2465]: This looks complicated:
```
def proces(...):
    while ...:
        for ....:
            proces(...)
```

but I don't think there is a way to simplify it

File: odoo/models.py
Comment (AdrienHorgnies) [inline odoo/models.py:2457]: This comment relates to a loop you've deleted.

File: odoo/models.py
Comment (AdrienHorgnies) [reply odoo/models.py:2465]: Indeed, and that would be a good reason to document it more. Especially the loops and recursive call. It's not clear what part of the problem each of them is handling.

What's the responsibility of the `while` loop compared to the `for` loop ?

What's the responsibility of the recursive call compared to the encompassing block ?

I'm having a hard time seeing when will the `while` loop do several laps, as it (erroneously) seem to me that the `for` loop and recursive call are doing everything.

**Review by AdrienHorgnies — CHANGES_REQUESTED** [review]
A bit hard to understand. It needs comments / documentation.

File: odoo/models.py
Comment (yelizariev) [reply odoo/models.py:2465]: I'll try to make some drawing to explain. It's not clear enough for me too :D -- I made the tests first and then made this code almost by intuition

File: odoo/models.py
Comment (yelizariev) [reply odoo/models.py:2483]: The idea is to don't change original code, which usues fname (see following lines of code)

File: odoo/models.py
Comment (yelizariev) [reply odoo/models.py:2518]: A possible option to don't use both while-loop and for-loop is something like this;

```
def process(...):
    if field is not related:
        return
    for ....:
        process(...)
        if it's last element in for-loop:
           proces(...)
        else:
            do join
```

Not sure if it's easy to read -- we get two places to call process instead of one

---

I've added comment to current implementation. Please check if it's good enough now

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2456]: I don't quite understand meaning of `(no duplicates)` remark. @rco-odoo ?

File: odoo/models.py
Comment (AdrienHorgnies) [reply odoo/models.py:2542]: Well, you almost changed the whole function, might as well modify this bit too.

File: odoo/models.py
Comment (AdrienHorgnies) [inline odoo/models.py:2450]: TODO

**Review by AdrienHorgnies — CHANGES_REQUESTED** [review]
Very nice documentation, much clearer ! I definitively would get rid of the `while` loop. As I suspected, it essentially does the same thing as the recursive calls, albeit only for the last field.

File: odoo/models.py
Comment (yelizariev) [reply odoo/models.py:2524]: So, it's
```
def process(...):
    if field is not related:
        return
    for join_fname in field.related[:-1]::
        process(...)
        do join
    call process() for last field
```
:thinking: looks better :+1: 
I'll try

Comment (yelizariev) [reply odoo/models.py]: Done! Thanks!

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2481]: image

File: odoo/models.py
Comment (yelizariev) [reply odoo/models.py:2456]: Is it about the fact we cannot do inherits for same table twice?

File: odoo/fields.py
Comment (yelizariev) [inline odoo/fields.py:602]: :point_up:

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2311]: :point_up: 

My previous variant was as following:

            # for non-stored related field we do a quick check here and deep check in _inherits_join_calc
           assert (gb_field.store or gb_field.related and gb_field.related_field.store) and gb_field.column_type, "Fields in 'groupby' must be regular database-persisted fields, or function fields with store=True, or related field without one2many element in the chain"

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2452]: :point_up:

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2456]: :point_up:

File: odoo/models.py
Comment (yelizariev) [inline odoo/models.py:2551]: :point_up:

**Review by yelizariev — COMMENTED** [review]
@Elkasitu please check my adaptation of this PR with your recent update https://github.com/odoo/odoo/pull/68958

**Review by AdrienHorgnies — APPROVED** [review]
I'm approving the PR despite that I wouldn't want to merge a TODO comment, as I'm off for two weeks and don't want to block the PR for this. Well, anyway, robodoo doesn't care about GitHub approvals, but that way it's clearer.

So @rco-odoo or whoever will merge, please check `# TODO: method should be renamed to _related_join_calc` before merging.

**Review by Elkasitu — COMMENTED** [review]
Ok for the changes relating to #68958

**Comment by yelizariev** [comment]
I've fixed conflicts related to recent update in master 
https://github.com/odoo/odoo/commit/1abe965b59a6acc24b93c25c07c2f9151af4924a 
https://github.com/odoo/odoo/commit/1ca2c7e457a688b2288536416c0a7350172e0cda (?)

**Comment by yelizariev** [comment]
Will we have it in Odoo 15 ?

**Comment by randall-vx** [comment]
@moylop260 @nhomar @hugho-ad

**Comment by william-andre** [comment]
> Will we have it in Odoo 15 ?

Will we have it in Odoo 16?

**Comment by yelizariev** [comment]
"let's talk about it after OXP" (c)

**Comment by h4818** [comment]
> > Will we have it in Odoo 15 ?
> 
> Will we have it in Odoo 16?

Will we have it in Odoo 17?

**Comment by yelizariev** [comment]
I made a new PR because runbot is more strict now about branch naming https://github.com/odoo/odoo/pull/104990

**Pull request closed without merging**
